### PR TITLE
Receive statistics from cloned actors for non-strikes

### PR DESCRIPTION
**Pull request by CarlosFdez**
Closes https://github.com/foundryvtt/pf2e/issues/12521

I would feel more comfortable having this merged if some people volunteer for extensive QA.

### CODE REVIEW

File: src/module/actor/roll-context/base.ts [hunk 1]
```diff
@@ -319,27 +319,54 @@ abstract class RollContext<
...
+        }
+
+        // If this actor isn't actually a clone, return the normal statistic
+        if (!unresolvedStatistic || !clonedActor || unresolvedRoller?.actor === clonedActor) {
+            return unresolvedStatistic ?? null;
+        }
+
+        // Note: we may need to eventually handle heirarchies by finding the first available parent and then applying diffs
+        // For now we simply return the original if we detect a deviation
+        const clonedStatistic = clonedActor.getStatistic(unresolvedStatistic.slug) ?? null;
```
Comment (stwlam) [inline src/module/actor/roll-context/base.ts:351]:
```suggestion
        const clonedStatistic = clonedActor.getStatistic(unresolvedStatistic.slug);
```

File: src/module/actor/roll-context/base.ts [hunk 2]
```diff
@@ -319,27 +319,54 @@ abstract class RollContext<
 
     #getClonedStatistic(clonedActor: ActorPF2e | null): Statistic | StrikeData | null {
         const unresolvedRoller = this.unresolved[this.rollerRole];
+        const unresolvedStatistic = unresolvedRoller?.statistic;
```
Comment (stwlam) [inline src/module/actor/roll-context/base.ts:322]:
```suggestion
        const unresolvedStatistic = unresolvedRoller?.statistic ?? null;
```

File: src/module/actor/roll-context/base.ts [hunk 3]
```diff
@@ -319,27 +319,54 @@ abstract class RollContext<
...
+                    action.item.isOfType("weapon") &&
+                    unclonedItem.isMelee === action.item.isMelee
+                );
+            });
+            return matchingStrike ?? unresolvedStatistic ?? null;
+        }
+
+        // If this actor isn't actually a clone, return the normal statistic
+        if (!unresolvedStatistic || !clonedActor || unresolvedRoller?.actor === clonedActor) {
+            return unresolvedStatistic ?? null;
```
Comment (stwlam) [inline src/module/actor/roll-context/base.ts:346]:
```suggestion
            return unresolvedStatistic;
```

File: src/module/actor/roll-context/base.ts [hunk 4]
```diff
@@ -319,27 +319,54 @@ abstract class RollContext<
...
+        // Note: we may need to eventually handle heirarchies by finding the first available parent and then applying diffs
+        // For now we simply return the original if we detect a deviation
+        const clonedStatistic = clonedActor.getStatistic(unresolvedStatistic.slug) ?? null;
+        const initialHeirarchy = this.#getStatisticHierarchy(unresolvedStatistic);
+        const clonedHeirarchy = this.#getStatisticHierarchy(clonedStatistic);
+        if (!R.isDeepEqual(initialHeirarchy, clonedHeirarchy)) {
+            return unresolvedStatistic ?? null;
+        }
+
+        return clonedStatistic ?? unresolvedStatistic ?? null;
```
Comment (stwlam) [inline src/module/actor/roll-context/base.ts:358]:
```suggestion
        return clonedStatistic ?? unresolvedStatistic;
```

File: src/module/actor/roll-context/base.ts [hunk 5]
```diff
@@ -319,27 +319,54 @@ abstract class RollContext<
...
+            return unresolvedStatistic ?? null;
+        }
+
+        // Note: we may need to eventually handle heirarchies by finding the first available parent and then applying diffs
+        // For now we simply return the original if we detect a deviation
+        const clonedStatistic = clonedActor.getStatistic(unresolvedStatistic.slug) ?? null;
+        const initialHeirarchy = this.#getStatisticHierarchy(unresolvedStatistic);
+        const clonedHeirarchy = this.#getStatisticHierarchy(clonedStatistic);
+        if (!R.isDeepEqual(initialHeirarchy, clonedHeirarchy)) {
+            return unresolvedStatistic ?? null;
```
Comment (stwlam) [inline src/module/actor/roll-context/base.ts:355]:
```suggestion
            return unresolvedStatistic;
```

File: src/module/actor/roll-context/base.ts [hunk 6]
```diff
@@ -319,27 +319,54 @@ abstract class RollContext<
...
+                if (unclonedItem.isOfType("melee") && action.item.isOfType("melee")) return true;
+
+                // Discriminate between melee/thrown usages by checking that both are either melee or ranged
+                return (
+                    unclonedItem.isOfType("weapon") &&
+                    action.item.isOfType("weapon") &&
+                    unclonedItem.isMelee === action.item.isMelee
+                );
+            });
+            return matchingStrike ?? unresolvedStatistic ?? null;
```
Comment (stwlam) [inline src/module/actor/roll-context/base.ts:341]:
```suggestion
            return matchingStrike ?? unresolvedStatistic;
```

File: src/module/actor/roll-context/base.ts (hunk 1 above)
Comment (CarlosFdez) [reply src/module/actor/roll-context/base.ts:351]: This one was silly. I forgot to check if we needed ?? null after I was no longer doing clonedActor?.getStatistic().

**Pull request merged**